blog/content/post/2020-12-08-IOMMU.md

```diff
@@ -0,0 +1,87 @@
+---
+title: What is IOMMU and how it can be used?
+abstract: 'Introductory blog post on how IOMMU works.
+          In this article you can read what IOMMU is
+          and find out if its use may be beneficial for you.'
+cover: /covers/IOMMU.png
+author: marek.kasiewicz
+layout: post
+published: true
+date: 2020-12-08
+archives: "2020"
+
+tags:
+  - Firmware
+  - IOMMU
+  - Security
+categories:
+  - Firmware
+  - Security
+
+---
+
+## Introduction
+
+The I/O memory management unit (IOMMU) is a type of memory management unit (MMU)
+that connects a Direct Memory Access (DMA) capable expansion bus to the main memory.
+It extends the system architecture by adding support for the virtualization of
+memory addresses used by peripheral devices. Additionally, it provides memory
+isolation and protection by enabling system software to control which areas of
+physical memory an I/O device may access. It also helps filter and remap interrupts
+from peripheral devices. 
+
+Advantages of IOMMU usage:
+
+  - One single contiguous virtual memory region can be mapped to multiple non-contiguous
+  physical memory regions. IOMMU can make a non-contiguous memory region appear
+  contiguous to a device (scatter/gather). Scatter/gather optimizes streaming DMA
+  performance for the I/O device.
+  - Memory isolation and protection: device can only access memory regions that are
+  mapped for it. Hence faulty and/or malicious devices can't corrupt memory.
+  - Memory isolation allows safe device assignment to a virtual machine without
+  compromising host and other guest OSes.
+  - IOMMU enables 32-bit DMA capable devices to access to > 4GB memory.
+  - Support hardware interrupt remapping. It extends limited hardware interrupts
+  to software interrupts. Primary uses are the interrupt isolation and translation
+  between interrupt domains (E.g. ioapic vs x2apic on x86)
+
+Disadvantages:
+
+  - Latency in dynamic DMA mapping, translation overhead penalty.
+  - Host software must maintain in-memory data structures for use by the IOMMU
```

Comment:
```suggestion
  - Host software must be maintained in-memory data structures for use by the IOMMU
```

Reply:
Software is a thing that has to maintain structures, not a thing that has to be maintained.